src/prediction.jl

```diff
@@ -1,37 +1,79 @@
-non_na_mean(vv) = mean(vv[.!(isnan.(vv))])
+########################################
+# 0. Hierarchical Structures and Types #
+########################################
 
-function fix_metadata_colnames!(longdata_df::DataFrame)
+abstract type Prediction end
+struct Classification <: Prediction end
+struct Regression <: Prediction end
 
-    ## This function was created to fix small inconsistencies and typos on the longdata table.
-    ## As original data cleanup and wrangling progresses, it will be altered and ultimately deprecated.
-    ## Comments are provided to help review transformations
+abstract type Predictor end
+abstract type UnivariatePredictor <: Resonance.Predictor end
+abstract type MultivariatePredictor <: Resonance.Predictor end
 
-    # 1. Rename "[Collinsella]_massiliensis" to "Collinsella_massiliensis"
-    longdata_df[longdata_df.variable .== "[Collinsella]_massiliensis", :variable] .= "Collinsella_massiliensis"
+mutable struct UnivariateRandomForestClassifier <: Resonance.UnivariatePredictor
+    name::String
+    original_data::DataFrame
+    inputs_outputs::Tuple{DataFrame, CategoricalArrays.CategoricalVector{Bool, UInt32, Bool, CategoricalArrays.CategoricalValue{Bool, UInt32}, Union{}}}
+    n_splits::Int64
+    dataset_partitions::Vector{Tuple{Vector{Int64}, Vector{Int64}}}
```

Comment:
Regarding constraints again - why couldn't these be a range for example, or another way of indexing?

Reply:
They cant be a range because they are shuffled indexes. Is there any other way of indexing that's simpler than a vector of integers for train, another for test? The only thing I can think of is a Bool[] and I think it's just as simple for the use cases we have, except that, for exploratory purposes, you lose the ability to look at a certain row of the input and easily find that same row on the original dataframe.